Added the ability to close a directory node

DrDeano opened the pull request:
Closes #270

SamTebbs33 commented on src/kernel/filesystem/vfs.zig:
```diff
@@ -87,13 +87,13 @@ pub const FileSystem = struct {
 
     ///
-    /// Close an open file, performing any last operations required to save data etc.
+    /// Close an open file or directory, performing any last operations required to save data etc.
```
Maybe change this to say "an open node" since any node, not just directories and files, can be closed with these changes.

SamTebbs33 commented on src/kernel/filesystem/vfs.zig:
```diff
@@ -191,7 +191,7 @@ pub const FileNode = struct {
     /// See the documentation for FileSystem.Close
     pub fn close(self: *const FileNode) void {
-        return self.fs.close(self.fs, self);
+        return self.fs.close(self.fs, @ptrCast(*const Node, self));
```
Instead of casting it would make more sense if this function becomes part of the `Node` struct.

DrDeano replied:
This wouldn't work as you can't close a symlink. Plus openFile and openDir return FileNode and DirNode so there will need to be a cast at some point.

SamTebbs33 replied:
I reckon we could close a symlink, after all it's going to be kept around in memory by the FS and we'll want to get rid of it from memory once it isn't used. openFile and openDir do indeed return specialised nodes, but when converting a `Node` to a `DirNode` or `FileNode` we switch on them rather than cast since it's much safer and allows us to detect invalid nodes being passed.

DrDeano replied:
How would you do back from a `FileNode` or `DirNode` to a `Node`?

SamTebbs33 replied:
Well you wouldn't have to if the function is part of `Node` and takes a `*Node` or `*const Node` as `self`.

DrDeano replied:
I had a look, but changes a lot of types from FileNode and DirNode to just Node and mean that I will need to change the mount type to Node instead of DirNode. I'm not sure how to do this. Could you show me da way sensei.
Was thinking of having 2 close functions, one for FileNode and one for DirNode.

SamTebbs33 replied:
Oh I see that the problem is that with this a close function that takes a `Node` we can't do
```
var file = vfs.openFile(...)
file.close()
```

A solution is to have a `close` function in in `FIleNode` and `DirNode` like you do now, then have a `close` function for each of those in `FileSystem` (i.e. `closeFile`, `closeDir`). This avoids the casting that could be dodgy. I'm not sure how to tackle closing symlinks, that could be tended to later on.

SamTebbs33 commented on src/kernel/filesystem/vfs.zig:
```diff
@@ -218,6 +218,17 @@ pub const DirNode = struct {
     }
+
+    /// See the documentation for FileSystem.Close
+    pub fn close(self: *const DirNode) void {
```
As with `FileNode.close`, we should make the close function be part of the `Node` struct instead.

SamTebbs33 reviewed (approved):
Noice

DrDeano committed:
Added the ability to close a directory node

Updated docs
Closes #270


Update doc comment


Moved close to closeFile and closeDir in VFS


Moved back to @ptrCast()


Always forget to format


Added TODO comment
